src/GitHub.Api/Tasks/DownloadTask.cs

```diff
@@ -0,0 +1,316 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net;
+using System.Text;
+using System.Threading;
+
+namespace GitHub.Unity
+{
+    public class Utils
+    {
+        public static bool Copy(Stream source, Stream destination, int chunkSize)
+        {
+            return Copy(source, destination, chunkSize, 0, null, 1000);
+        }
+
+        public static bool Copy(Stream source, Stream destination, int chunkSize, long totalSize,
+            Func<long, long, bool> progress, int progressUpdateRate)
+        {
+            byte[] buffer = new byte[chunkSize];
+            int bytesRead = 0;
+            long totalRead = 0;
+            float averageSpeed = -1f;
+            float lastSpeed = 0f;
+            float smoothing = 0.005f;
+            long readLastSecond = 0;
+            long timeToFinish = 0;
+            Stopwatch watch = null;
+            bool success = true;
+
+            bool trackProgress = totalSize > 0 && progress != null;
+            if (trackProgress)
+                watch = new Stopwatch();
+
+            do
+            {
+                if (trackProgress)
+                    watch.Start();
+
+                bytesRead = source.Read(buffer, 0, chunkSize);
+
+                if (trackProgress)
+                    watch.Stop();
+
+                totalRead += bytesRead;
+
+                if (bytesRead > 0)
+                {
+                    destination.Write(buffer, 0, bytesRead);
+                    if (trackProgress)
```

Comment:
This if-statement is just figuring out how long it took to download, correct?

Reply:
That code is attempting to figure out the progress of the download in real time.
But my tests of the progress reporting has shown me it's less accurate that one would like.
I'm not using it now, but I will record an issue to address it because I would like to use it soon.